tests/run/TupleFlatMap.scala

```diff
@@ -0,0 +1,19 @@
+@main def Test = {
+
+  println(
+    Tuple().flatMap[EmptyTuple, [t] =>> (t, t)]([t] => (x: t) => (x, x))
+  )
+
+  println(
+    (1, 2, "x", "d").flatMap[(Int, Int, String, String), [t] =>> (t, t)]([t] => (x: t) => (x, x))
+  )
+
+  println(
+    (1, "x").flatMap[(Int, String), [t] =>> EmptyTuple]([t] => (x: t) => Tuple())
```

Comment:
The use of this API seems clunky. Do we need the explicit type parameters?

Reply:
I checked and it turns out that the explicit type params are needed. Though this PR shouldn't be merged yet.
Examples 3 and 4 result in a compilation error, but the first two examples end with compiler crash like:
```
Fatal compiler crash when compiling: tests/run/TupleFlatMap.scala:
assertion failed: orphan param: t, hash of binder = 1998366311, tree = <empty>, type = TypeVar(TypeParamRef(F) -> HKTypeLambda(List(T2), List(TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any))), AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Tuple2),List(TypeParamRef(t), TypeParamRef(T2)))))
```
I'll take a look at it and try to fix it.

Reply:
Minimized it in #13286